// app/melody/auth/api/rest/login.js
import createDebugger       from 'evoke-me/debug/api/create'

import config               from '../config'
import errorMap             from '../../../manager/api/errorMap'


const who = [config.namespace, config.entity, `api`, `rest`, `login`]
const debug = createDebugger(who.join(`:`))

export default async (payload, options = {}) => {
  try {
    debug(payload, options)
    const {
      mobileNo,
      password,
    } = payload
    const {
      host = `https://api.melodycity.cn/v1`
    } = options


    const response = await fetch(`${host}/user/login?mobileNo=${mobileNo}&password=${password}`)
    const json = await response.json()

    const { code, message, data: { sessionKey }} = json

    if (code === 0) {
      return {
        result: {
          message,
          sessionKey,
        },
        errors: [],
      }
    }

    return {
      result: null,
      errors: [{
        type: config.namespace,
        code: errorMap.ERROR_CODE_NOT_ZERO,
        message,
        props: {
          code,
        }
      }],
    }
  } catch (err) {
    debug(`ERROR`, err)

    return {
      result: null,
      errors: [{
        type: config.namespace,
        code: errorMap.UNKNOWN,
        message: `Unknown error`,
      }]
    }
  }
}
